[Security Solution] Add event-based telemetry for prebuilt rule upgrade API

jkelas opened the pull request:
**Partially resolves: #140369**

## Summary

This is another PR from of a series of PRs I am planning to create to cover the requirements in the #140369 ticket. 

The requirement covered in this ticket is req. #6: "Events for performing update (EBT backend)" and req. #7 "Missing base versions".

I am adding sending telemetry events in handling of rule update request. Each rule updated will send its own event with information about:
- ruleId
- ruleName
- if missing base version
- final result of the update
- updated fields (with breakdown per conflict type). 

I tried to make the changes as little invasive as possible, and decided to create a separate file, `update_rule_telemetry.ts`, where the logic of building the events and sending them is encapsulated.

### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.




<!--ONMERGE {"backportTargets":["8.18","8.19","9.0","9.1"]} ONMERGE-->

Commit:
Add telemetry for rule update

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/338017)
* Commit: 921557835d077abc41863cc0bd4bb3d82259e503

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/338017#01993447-4ed5-4f3d-be09-5dca5ed53f12)

cc @jkelas

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338017","number":338017,"commit":"921557835d077abc41863cc0bd4bb3d82259e503"}],"number":338017}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/338056)
* Commit: b052147174c39d3580bd5939f30bce0c1da902ae

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/6e4af9dd4e141e308fb465be8ddb2e0b647c53c8) | [after](https://github.com/elastic/kibana/commit/b052147174c39d3580bd5939f30bce0c1da902ae) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 131 | 132 | +1 |

#### Saved Objects .kibana field count
> Every field in each saved object type adds overhead to Elasticsearch. Kibana needs to keep the total field count below Elasticsearch's default limit of 1000 fields. Only specify field mappings for the fields you wish to search on or query. See https://www.elastic.co/guide/en/kibana/master/saved-objects-service.html#_mappings


| id | [before](https://github.com/elastic/kibana/commit/6e4af9dd4e141e308fb465be8ddb2e0b647c53c8) | [after](https://github.com/elastic/kibana/commit/b052147174c39d3580bd5939f30bce0c1da902ae) | diff |
| --- | --- | --- | --- |
| `_data_stream_timestamp` | 1 | - | -1 |
| `_doc_count` | 1 | - | -1 |
| `_ignored_source` | 1 | - | -1 |
| `_index_mode` | 1 | - | -1 |
| `_inference_fields` | 1 | - | -1 |
| `_tier` | 1 | - | -1 |
| `apm-custom-dashboards` | 5 | - | -5 |
| `apm-server-schema` | 2 | - | -2 |
| `apm-service-group` | 5 | - | -5 |
| `application_usage_daily` | 2 | - | -2 |
| `config` | 2 | - | -2 |
| `config-global` | 2 | - | -2 |
| `coreMigrationVersion` | 1 | - | -1 |
| `created_at` | 1 | - | -1 |
| `created_by` | 1 | - | -1 |
| `entity-definition` | 9 | - | -9 |
| `entity-discovery-api-key` | 2 | - | -2 |
| `event_loop_delays_daily` | 2 | - | -2 |
| `favorites` | 4 | - | -4 |
| `file` | 11 | - | -11 |
| `file-upload-usage-collection-telemetry` | 3 | - | -3 |
| `fileShare` | 5 | - | -5 |
| `infra-custom-dashboards` | 4 | - | -4 |
| `infrastructure-monitoring-log-view` | 2 | - | -2 |
| `intercept_trigger_record` | 5 | - | -5 |
| `legacy-url-alias` | 7 | - | -7 |
| `managed` | 1 | - | -1 |
| `ml-job` | 6 | - | -6 |
| `ml-module` | 13 | - | -13 |
| `ml-trained-model` | 7 | - | -7 |
| `monitoring-telemetry` | 2 | - | -2 |
| `namespace` | 1 | - | -1 |
| `namespaces` | 1 | - | -1 |
| `observability-onboarding-state` | 2 | - | -2 |
| `originId` | 1 | - | -1 |
| `product-doc-install-status` | 7 | - | -7 |
| `references` | 4 | - | -4 |
| `sample-data-telemetry` | 3 | - | -3 |
| `security-ai-prompt` | 8 | - | -8 |
| `slo` | 11 | - | -11 |
| `space` | 5 | - | -5 |
| `synthetics-monitor` | 34 | - | -34 |
| `synthetics-monitor-multi-space` | 34 | - | -34 |
| `tag` | 4 | - | -4 |
| `type` | 1 | - | -1 |
| `typeMigrationVersion` | 1 | - | -1 |
| `ui-metric` | 2 | - | -2 |
| `updated_at` | 1 | - | -1 |
| `updated_by` | 1 | - | -1 |
| `upgrade-assistant-ml-upgrade-operation` | 3 | - | -3 |
| `upgrade-assistant-reindex-operation` | 3 | - | -3 |
| `uptime-synthetics-api-key` | 2 | - | -2 |
| `url` | 5 | - | -5 |
| `usage-counters` | 2 | - | -2 |
| total |  |  | -246 |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/6e4af9dd4e141e308fb465be8ddb2e0b647c53c8) | [after](https://github.com/elastic/kibana/commit/b052147174c39d3580bd5939f30bce0c1da902ae) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 200 | 201 | +1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/6e4af9dd4e141e308fb465be8ddb2e0b647c53c8) | [after](https://github.com/elastic/kibana/commit/b052147174c39d3580bd5939f30bce0c1da902ae) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 677 | 678 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/6e4af9dd4e141e308fb465be8ddb2e0b647c53c8) | [after](https://github.com/elastic/kibana/commit/b052147174c39d3580bd5939f30bce0c1da902ae) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 779 | 780 | +1 |


</details>


### History
* :broken_heart: [Build #338017](https://buildkite.com/elastic/kibana-pull-request/builds/338017) failed 921557835d077abc41863cc0bd4bb3d82259e503

cc @jkelas

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338056","number":338056,"commit":"b052147174c39d3580bd5939f30bce0c1da902ae"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338017","number":338017,"commit":"921557835d077abc41863cc0bd4bb3d82259e503"}],"number":338056}
buildkite-pr-comment-->

szaffarano reviewed (approved):
LGTM

maximpn reviewed (changes requested):
@jkelas Thanks for keeping the changes minimal and focused 🙏 

I left comments to further improve the implementation taking into account cohesion and our future plans on `upgrade/_perform` API endpoint refactoring.

jkelas commented on x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/event_based/events.ts:
```diff
@@ -27,6 +27,94 @@ import type {
 
 import { SIEM_MIGRATIONS_EVENTS } from './events/siem_migrations';
 
+export const DETECTION_RULE_UPDATE_EVENT: EventTypeOpts<{
```
Ok, I will add a comment

jkelas commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/update_rule_telemetry.ts:
```diff
@@ -0,0 +1,159 @@
+  const updatedFieldsWithNoConflicts: string[] = [];
+
+  for (const [fieldName, diff] of Object.entries(calculatedRuleDiff)) {
+    if (fieldName !== 'version') {
```
Ok, will improve.

Commit:
Fixes from review - part 1

Commit:
Fixes from review - part 2

jkelas commented on x-pack/solutions/security/plugins/security_solution/server/request_context_factory.ts:
```diff
@@ -144,6 +144,8 @@ export class RequestContextFactory implements IRequestContextFactory {
     return {
       core: coreContext,
 
+      analytics: core.analytics,
```
As agreed offline, moving to CoreRequestHandlerContext will be covered as a follow-up PR, if the Kibana core approves (contacted offline).

Commit:
Merge branch 'main' into 140369_telemetry_6_ebt_server_2

Commit:
Remove export

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/339705)
* Commit: 0aef0e851a8fa26daae84bd2016ffb96c8315f45

### Failed CI Steps
* [Serverless Investigations - Security Solution Cypress Tests #6](https://buildkite.com/elastic/kibana-pull-request/builds/339705#019951df-7a94-4288-a831-3a4b257b7b7a)

### History
* :broken_heart: [Build #338938](https://buildkite.com/elastic/kibana-pull-request/builds/338938) failed bb3bb3e3ca383ab9c5e381f7a92f8c672e2f9561
* :green_heart: [Build #338056](https://buildkite.com/elastic/kibana-pull-request/builds/338056) succeeded b052147174c39d3580bd5939f30bce0c1da902ae
* :broken_heart: [Build #338017](https://buildkite.com/elastic/kibana-pull-request/builds/338017) failed 921557835d077abc41863cc0bd4bb3d82259e503

cc @jkelas

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339705","number":339705,"commit":"0aef0e851a8fa26daae84bd2016ffb96c8315f45"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338938","number":338938,"commit":"bb3bb3e3ca383ab9c5e381f7a92f8c672e2f9561"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338056","number":338056,"commit":"b052147174c39d3580bd5939f30bce0c1da902ae"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338017","number":338017,"commit":"921557835d077abc41863cc0bd4bb3d82259e503"}],"number":339705}
buildkite-pr-comment-->

maximpn reviewed (approved):
@jkelas Thanks for addressing my comments 👍 

I've left a few non-critical comments and hope you'll find time to address them.

Commit:
More improvements from review

jkelas commented:
> @jkelas Thanks for addressing my comments 👍
> 
> I've left a few non-critical comments and hope you'll find time to address them.

Thank you Maxim, I addressed all your comments.

Commit:
Merge branch 'main' into 140369_telemetry_6_ebt_server_2

jkelas commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/235315">](https://github.com/elastic/kibana/pull/235315)|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/235317">](https://github.com/elastic/kibana/pull/235317)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/235318">](https://github.com/elastic/kibana/pull/235318)|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/235319">](https://github.com/elastic/kibana/pull/235319)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.2 -->
